azurerm/internal/services/mysql/mysql_server_resource.go

```diff
@@ -24,6 +24,10 @@ import (
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
 )
 
+const (
+	mySQLServerResourceName = "azurerm_mysql_server"
+)
```

Comment:
presumably we'll also need to lock in the create/update/delete here too?

Reply:
I assume we do not need locks in postgresql server.

This key resource has a kind of weird behaviour - it is a separated resource from the server by service design, but it is a singleton, you could only have one key at one time. If you are trying to create two keys (with different name) for one server, after the second key is created on the service side, the first key will be automatically removed without any notice.

Therefore we need locks in the server key resource to ensure that if the user is not using this resource correctly (by say create two keys for one server), the order of key creation is predictable - this will always give the user a diff after apply.

The server does not have this kind of constraint, therefore I suppose the server resource does not need locks.

Reply:
so what happens if you try and decrypt the server and delete it simultaneously? a conflict will likely occur somewhere

Reply:
By decrypting the server, I assume we are deleting the server key  - since we locked the server in the delete function of server key resource, we could not decrypt the server and delete it simultaneously - during the decrypting, the server is locked and deletion should hold on.

Reply:
can we test this to confirm?

Reply:
@ArcturusZhang any update here?

Reply:
Tests have been done per requested.